### ISSUE TITLE
Exception: timeout when polling for transaction

### ISSUE THREAD

**Issue by drandreaskrueger** [raiden-network/raiden#1058]
after downgrading [solc to v0.4.16](https://github.com/raiden-network/raiden/issues/1057#issuecomment-331897269) I see no more wall of warnings, but this:

```
Welcome to Raiden, version 0.1.0+git.r72845c7a!
ERROR:raiden.network.upnpsock	could not query your lanaddr reported=
INFO:raiden.network.sockfactory	Network port opened internal=0.0.0.0:40001 method=stun external=my.EXTERNAL.ip.address:40001
Checking if the ethereum node is synchronized
Warning: This is a pre-release compiler version, please do not use it in production.
Warning: This is a pre-release compiler version, please do not use it in production.
Warning: This is a pre-release compiler version, please do not use it in production.
Warning: This is a pre-release compiler version, please do not use it in production.
Warning: This is a pre-release compiler version, please do not use it in production.
```
and then after ~5 minutes:
```
Traceback (most recent call last):
  File "PATH/raiden-network/raiden/env/raidenpython/local/lib/python2.7/site-packages/gevent/greenlet.py", line 534, in run
    result = self._run(*self.args, **self.kwargs)
  File "PATH/raiden-network/raiden/raiden/network/discovery.py", line 92, in register
    self.discovery_proxy.register_endpoint(node_address, endpoint)
  File "PATH/raiden-network/raiden/raiden/network/rpc/client.py", line 580, in register_endpoint
    timeout=self.poll_timeout,
  File "PATH/raiden-network/raiden/env/raidenpython/local/lib/python2.7/site-packages/pyethapp/rpc_client.py", line 753, in poll
    raise Exception('timeout when polling for transaction')
Exception: timeout when polling for transaction
<Greenlet at 0x7ffb85374910: <bound method ContractDiscovery.register of <raiden.network.discovery.ContractDiscovery object at 0x7ffb842004d0>>('\x00=$\xb3\xb8^\x10\xf6\xa4i\xc1m\xa1\xb9\x05I\x1, 'my.EXTERNAL.ip.address', 40001)> failed with Exception

CRITICAL:raiden.raiden_service	Endpoint registry failed: Exception('timeout when polling for transaction',). Ethereum RPC API might be unreachable. 

```
with `my.EXTERNAL.ip.address` = edited out by me.

# Versions
```
source raiden/env/raidenpython/bin/activate; raiden version; parity --version; solc --version; deactivate

{
  "python_version": "2.7.13", 
  "python_implementation": "CPython", 
  "system": "Linux 64bit_ 4.9.0-3-amd64", 
  "raiden": "0.1.0+git.r72845c7a"
}

Parity
  version Parity/v1.8.0-unstable-684b14271-20170921/x86_64-linux-gnu/rustc1.20.0

solc
Version: 0.4.16-develop.2017.9.25+commit.d7661dd9.Linux.g++
```

**Comment by hackaugusto** [comment]
From the issues you have opened my guess is that you're on a private chain that doesn't have the smart contracts deployed.

Here is what you need to check:

- make sure parity / geth is running.
- make sure the client has the json rpc interface enabled, [parity](https://github.com/paritytech/parity/wiki/JSONRPC#enabled-apis-default) [geth](https://github.com/ethereum/go-ethereum#programatically-interfacing-geth-nodes).
- if you're running on a [testnet](http://raiden-network.readthedocs.io/en/stable/overview_and_guide.html#firing-it-up), make sure the node is fully synchronized.
- if you're running a private chain, make sure the required smart contracts are [deployed](https://github.com/raiden-network/raiden/blob/master/docs/smart_contracts.rst).
- make sure raiden is executed with all the proper flags for the above (`--eth-rpc-endpoint`, `--registry-contract-address`, and `--discovery-contract-address`, use `--help` for more details)

**Comment by drandreaskrueger** [comment]
> you're on a private chain that doesn't have the smart contracts deployed.

Not sure I understand. I have synced Ropsten - that is a public chain, right?


As parity did not work, I have now also tried the same with geth, following [your manual](http://raiden-network.readthedocs.io/en/stable/overview_and_guide.html#using-geth)

```
geth --testnet --fast --bootnodes "enode://20c9ad97c081d63397d7b685a412227a40e23c8bdc6688c6f37e97cfbc22d2b4d1db1510d8f61e6a8866ad7f0e17c02b14182d37ea7c3c8b9c2683aeb6b733a1@52.169.14.227:30303,enode://6ce05930c72abc632c58e2e4324f7c7ea478cec0ed4fa2528982cf34483094e9cbc9216e7aa349691242576d552a2a56aaeae426c5303ded677ce455ba1acd9d@13.84.180.240:30303"

raiden --keystore-path  ~/.ethereum/testnet/keystore --address 0xc... --password-file password.txt
```

and now I get these results:

```
Welcome to Raiden, version 0.1.0+git.r72845c7a!
connect: Connection timed out
connect: Connection timed out
ERROR:raiden.network.upnpsock	Error when connecting to uPnP provider exception_info=No UPnP device discovered
INFO:raiden.network.sockfactory	Network port opened internal=0.0.0.0:40001 method=stun external=my.EXTERNAL.ip.address:40001
Couldn't determine the network the ethereum node is connected to.
Because of this there is not way to determine the latest
block with an oracle, and the events from the ethereum
node cannot be trusted. Giving up.
```

with `my.EXTERNAL.ip.address` = edited out by me.

**Comment by drandreaskrueger** [comment]
Just to clarify, what I mean by with `my.EXTERNAL.ip.address`:

I have an internal ip address, 192.168.x.y - in our local LAN, which I get when I do:

    ip a | grep eth0 | grep inet
        192.168.x.y

I have an external ip address by which however my machine is NOT visible on the internet:

    curl ifconfig.co
        217.x.y.z
    
the latter one is shown where I replaced with `my.EXTERNAL.ip.address`

**Comment by LefterisJP** [comment]
Is ropsten fully synced? Did you check against https://ropsten.etherscan.io/ ?

Also can you also add some extra arguments to your geth when you start it?

We have them added in the docs, but since no new release has been made yet they are only in the [github version](https://github.com/raiden-network/raiden/blob/master/docs/overview_and_guide.rst#using-geth) of the docs.

Specifically add: `--rpc --rpcapi eth,net,web3,debug`

RPC should definitely be open. And the rpc apis should be specified as above.

**Comment by drandreaskrueger** [comment]
Alright, the first problem I have found, the `--rpc` is missing in your manual:

http://raiden-network.readthedocs.io/en/stable/overview_and_guide.html#using-geth

**Comment by LefterisJP** [comment]
> Alright, the first problem I have found, the --rpc is missing in your manual:

@drandreaskrueger As I said above:

We have them added in the docs, but since no new release has been made yet they are only in the github version of the docs. A new docs version is generated only with a release.

**Comment by drandreaskrueger** [comment]
aha

**Comment by drandreaskrueger** [comment]
... but then after ~5 minutes this again:

```
Traceback (most recent call last):
  File "PATH/raiden-network/raiden/env/raidenpython/local/lib/python2.7/site-packages/gevent/greenlet.py", line 534, in run
    result = self._run(*self.args, **self.kwargs)
  File "PATH/raiden-network/raiden/raiden/network/discovery.py", line 92, in register
    self.discovery_proxy.register_endpoint(node_address, endpoint)
  File "PATH/raiden-network/raiden/raiden/network/rpc/client.py", line 580, in register_endpoint
    timeout=self.poll_timeout,
  File "PATH/raiden-network/raiden/env/raidenpython/local/lib/python2.7/site-packages/pyethapp/rpc_client.py", line 753, in poll
    raise Exception('timeout when polling for transaction')
Exception: timeout when polling for transaction
<Greenlet at 0x7f9c55dc8910: <bound method ContractDiscovery.register of <raiden.network.discovery.ContractDiscovery object at 0x7f9c554d85d0>>('\xc6-\x9fP\x92s\x7f\xa7h\x95y\xe2y\xa3\xa1\xddUC\, 'my.EXTERNAL.ip.address', 40001)> failed with Exception

CRITICAL:raiden.raiden_service	Endpoint registry failed: Exception('timeout when polling for transaction',). Ethereum RPC API might be unreachable. 
```

**Comment by LefterisJP** [comment]
This could be a ropsten problem. Let me try myself too. If ropsten is under congestion/empty blocks attack and transactions don't get included then raiden could:

1. Send a transaction, in your case for registering your endpoint.
2. Wait for it to get mined and included in a block
3. Keep waiting .. keep waiting ...
4. Hit the waiting timeout and give you this exception.

**Comment by drandreaskrueger** [comment]
Thanks.

In your above comment:

> make sure raiden is executed with all the proper flags for the above (--eth-rpc-endpoint, --registry-contract-address, and --discovery-contract-address, use --help for more details)

what exactly do I have to specify additionally your [the manuals](http://raiden-network.readthedocs.io/en/stable/overview_and_guide.html#firing-it-up)

--eth-rpc-endpoint ___    == that is standard 8545, right?

and about the contracts:

--registry-contract-address ___  
--discovery-contract-address ___

?

Can't I use the Ropsten defaults there, i.e. not specify them at all?

**Comment by LefterisJP** [comment]
yeah no need to specify any of those if you are in ropsten and also unless you change geth's rpc endpoint the default is used

**Comment by drandreaskrueger** [comment]
> 3. Keep waiting .. keep waiting ...
> 4. Hit the waiting timeout and give you this exception.

ah, now that sounds probable. 

So ... how can I manually extend the timeout, to wait for Ropsten miners?

**Comment by drandreaskrueger** [comment]
Or shall I perhaps even switch on CPU mining myself?

**Comment by drandreaskrueger** [comment]
Idea: What about launching a second version of all this on Kovan? I am actually not interested in fixing Ropsten's problems, I just want to test Raiden ;-)

**Comment by drandreaskrueger** [comment]
What about this recurring error message:

    ERROR:raiden.network.upnpsock	Error when connecting to uPnP provider exception_info=No UPnP device discovered

?

**Comment by LefterisJP** [comment]
one note: Since this seems to habe become a kind of a back and forth Q&A we also have a developers chat: https://gitter.im/raiden-network/raiden

> What about this recurring error message:

Raiden tries to do NAT Traversal. At first it tries to see if it can use UPnP and if this does not work it tries STUN.

**Comment by drandreaskrueger** [comment]
> developers chat: https://gitter.im/raiden-network/raiden

thank you very much, good to know.

I do appreciate the logging and better formatting of issues here (over gitter), but good to know, I might jump into gitter to ask shorter questions.

Thanks a lot.

**Comment by drandreaskrueger** [comment]
Redone it. Now I have noticed this TX in geth:
```
INFO [09-26|11:38:02] Imported new chain segment               blocks=1 txs=1  mgas=0.042 elapsed=4.241ms  mgasps=9.813  number=1749565 hash=7c2879…acb24c
INFO [09-26|11:38:03] Submitted transaction                    fullhash=0x09ca83db2397bcc46bfde3a6551285ec71325603deba21eca4b9aaf65cd1a9ed recipient=0x826259ce4DCC2802C92780e3d79D43fF3CF3f151
```

And this time it actually shows up on etherscan:

https://ropsten.etherscan.io/tx/0x09ca83db2397bcc46bfde3a6551285ec71325603deba21eca4b9aaf65cd1a9ed
https://ropsten.etherscan.io/address/0xc62d9f5092737fa7689579e279a3a1dd5543fb9e

I had that before, but it was a bit inconsistent: Those transactions did never get confirmed, and just disappeared again from etherscan - after perhaps 30 minutes.

**Comment by drandreaskrueger** [comment]
What about giving up on Ropsten?

What are the alternatives?

**Comment by drandreaskrueger** [comment]
Do the Raiden contracts exist on Kovan and/or Rinkeby?  
If not, is it much work to deploy them there?

**Comment by LefterisJP** [comment]
I don't know. We only deploy on ropsten so far as we need a single place of deployment and since we can't support multiple testing beds. We may need to reconsider with ropsten acting like that.

I can confirm you proble. My transactions also don't get confirmed on ropsten today.

You are always free to deploy the contracts yourself in any network you want though. We even have a guide on how to do so: https://github.com/raiden-network/raiden/blob/master/docs/smart_contracts.rst

**Comment by pcppcp** [comment]
I think there's another DoS attack on Ropsten - someone with a lot of hashpower is mining empty blocks and txs are not coming through. It's going on and off for at least a week.

**Comment by drandreaskrueger** [comment]
Hah. I went back to `parity` one last time (after that long `geth` detour), and *hooray* transactions had gotten mined:  

https://ropsten.etherscan.io/address/0x003d24b3b85e10f6a469c16da1b90549131ddfe3

I suppose that is good now, right?

```
Welcome to Raiden, version 0.1.0+git.r72845c7a!
ERROR:raiden.network.upnpsock	could not query your lanaddr reported=
INFO:raiden.network.sockfactory	Network port opened internal=0.0.0.0:40001 method=stun external=my.EXTERNAL.ip.address:40001
Checking if the ethereum node is synchronized
Warning: This is a pre-release compiler version, please do not use it in production.
Warning: This is a pre-release compiler version, please do not use it in production.
Warning: This is a pre-release compiler version, please do not use it in production.
Warning: This is a pre-release compiler version, please do not use it in production.
Warning: This is a pre-release compiler version, please do not use it in production.
INFO:raiden.network.discovery	endpoint already registered node_address=003d24b3 host=my.EXTERNAL.ip.address port=40001
INFO:filelock	Lock 140073336651664 acquired on /home/akrueger/.raiden/0x003d24/.lock 
The Raiden API RPC server is now running at http://127.0.0.1:5001/.

See the Raiden documentation for all available endpoints at
http://raiden-network.readthedocs.io/en/stable/rest_api.html
```
VERY HAPPY. That all took a subjective eternity, to get here.

And I do have trust issues now: With Ropsten broken/attacked/minerless/whatever/... today, would I run into new problems that come neither from raiden nor parity, but actually are simply caused by a shaky Ropsten? (That transaction got confirmed yesterday, and only after dozens of attempts.). What to do?   
Why not also deploy onto Kovan&Rinkeby - and if you do that, it won't take you much time (but if I try that, I fear to lose another day with only that). Thanks for considering it.

And most of all: Thanks for all your help!

**Comment by hackaugusto** [comment]
Closing in favor of #1134
